### ISSUE TITLE
``typing.Any`` is not recognized as type

### ISSUE THREAD

**Issue by gresm** [microsoft/pyright#10856]
**Description**
The ``typing.Any`` object isn't treated by pyright as a type, even though it is:
```py
from typing import Any
print(isinstance(Any, type))  # -> True
```
It's not an implementation detail, python documetation [calls it a type](https://docs.python.org/3/library/typing.html#the-any-type).

**Where Pyright fails**

```python
from typing import Any

def foo(t: type): ...

foo(Any) # Argument of type "type[Any]" cannot be assigned to parameter "t" of type "type" in function "foo"  (reportArgumentType)
```
Expected no errors.

Here's more sophisticated example:

```py
from typing import Any, TypeVar

T = TypeVar("T")
def foo(t: type[T]) -> T: ...

reveal_type(foo(Any)) # Argument of type "type[Any]" cannot be assigned to parameter "t" of type "type[T@foo]" in function "foo"  (reportArgumentType)
# Type of "foo(Any)" is "Unknown"
```
Expected ``Any`` and no errors.

**Comment by InSyncWithFoo** [comment]
Related: #8973.

**Comment by Azureblade3808** [comment]
When an argument has type `type[...]`, it only accepts classes as its values.

For example, code like

```python
from typing import Any

def foo(t: type[Any]) -> None: ...

foo(int | str)
```

won't work, though we can agree that the union type `int | str` is also a type.

Reference
---------
https://typing.python.org/en/latest/spec/special-types.html#type

>   The value corresponding to type[C] must be an actual class object that’s a subtype of C, not a [special form](https://typing.python.org/en/latest/spec/glossary.html#term-special-form) or other kind of type.

**Comment by Azureblade3808** [comment]
> It's not an implementation detail, python documetation [calls it a type](https://docs.python.org/3/library/typing.html#the-any-type).

`Any` is indeed a type, but whether it's a class (i.e. an instance of `type`) is implementation detail.

I do think that if `type[...]` were renamed something like `Class[...]`, there would be less misunderstanding.

**Comment by gresm** [comment]
@Azureblade3808 That's partially true, but not fully:
> When an argument has type type[...], it only accepts classes as its values.

One exception to this rule is type Union (eg: ``type[int | str]``), which unfortunately is your counterexample (same document as you are referencing, just a bit below).

Other case is ``type[Any]``. Said [reference](https://typing.python.org/en/latest/spec/special-types.html#type) (at the bottom) states it as a valid annotation and equivalent to plain ``type``.

It's also mentioned in passing that ``type[T]`` is also valid, but it's not elaborated on (only mentioned that the details are provided in some relevant document).

**Comment by Azureblade3808** [comment]
> One exception to this rule is type Union (eg: type[int | str]), which unfortunately is your counterexample (same document as you are referencing, just a bit below).

I'm not sure if I read it correctly. I know about type unions, and I believe that what I said also applies to them. Maybe you can share some code for me to better understand how there would be an exception?

**Comment by gresm** [comment]
I think I misunderstood what you meant with ``int | str`` being invalid. I was saying that ``type[int | str]`` is a valid annotation and from what I see now, you were saying that a variable ``a: type[Any]`` / ``a: type`` (as they are equivalent) cannot be equal to ``a = int | str`` as that's an object and not type, which is true, but I'm unsure how it is related to the issue...

**Comment by Azureblade3808** [comment]
1. There is no guarantee that the `Any` object should be a class (i.e. an instance of `type`), so type checker will treat it as a plain object.
2. Variables with type of `type[...]` only accept class objects as their values.
3. Because of 1 and 2, `Any` cannot be assigned to a variable with type of `type[...]`.

I talked about `int | str` since it's also a type but is obviously not a class, and thought it might better explain the situation that a type may not always be assignable to a variable with type of `type[...]`.

**Comment by gresm** [comment]
@Azureblade3808 Okay, I understand, but have doubts.
> There is no guarantee that the Any object should be a class (i.e. an instance of type), so type checker will treat it as a plain object.

``Any`` per documentation is defined as a type. Distinction between classes and types is pre-python2.2 anachronism. What guarantee ``Any`` lacks that I'm failing to see?

Though, I should note that ``Any`` is not instantiable, but this doesn't deprive it the status of a type.

**Comment by Azureblade3808** [comment]
The meaning of "type" in typing system and in runtime is not the same.

A type in typing system can be a class, a generic alias (e.g. `list[str]`), a special form (e.g. `Any`), a forward reference, or perhaps some other stuff I don't recall. `Any` is a type, but not a class.

**Comment by erictraut** [comment]
Yes, @Azureblade3808 is correct. `Any` is a special form. It is a "type form" in the type system and describes a type, but it is not itself assignable to `type` (which itself is a type form that specifically describes the set of all values that are instances of the `type` class or subclasses thereof). I realize the nomenclature can get confusing. You might find it useful to read the draft [PEP 747](https://peps.python.org/pep-0747/), which attempts to clear up some of the confusion and extend the type system so it's possible to annotate a function that accepts any type form including `Any`.

Pyright has provisional support for PEP 747 if you want to play with it. You'll need to set `enableExperimentalFeatures` to true in your config.

I'm going to close this issue because I think pyright is working correctly here.
